platform/linux-generic/odp_ipsec.c

```diff
@@ -565,18 +593,40 @@ static int ipsec_in_esp(odp_packet_t *pkt,
 	param->auth_iv_ptr = state->iv;
 
 	state->esp.aad.spi = esp.spi;
-	state->esp.aad.seq_no = esp.seq_no;
 	state->in.seq_no = odp_be_to_cpu_32(esp.seq_no);
+	if (ipsec_sa->esn)
+		ipsec_sa->in.seq64 = ipsec_compute_esn(ipsec_sa, state->in.seq_no);
```

Comment:
This is broken with multiple threads processing IPsec packets. You cannot store temporary per-thread per-packet data in a shared variable like that.

Reply:
Agreed, will fix it.

Reply:
You have not fixed the problem.

You cannot store temporary per-thread per-packet data to a variable (ipsec SA field) that is shared between threads. Multiple threads can be processing the SA at the same time and they would step on each other's toes.

You are still writing the calculated full sequence number of the currently processed packet in the SA. Adding spinlocking around the write does not fix the problem. Another thread may overwrite the field at any time once you release the spinlock.

The correct solution should be quite obvious. Take a look at how similar per-packet data is handled in the existing code.

Reply:
Thanks for correcting this. 
It seems the "ipsec_state_t" is used for the temporary storage of per-thread per-packet data, so I will move the "uint64_t seq64" from the SA data structure to "ipsec_state_t".

Reply:
Yes, that will fix the problem.